applications/FluidDynamicsApplication/custom_conditions/navier_stokes_wall_condition.cpp

```diff
@@ -63,6 +65,207 @@ void NavierStokesWallCondition<3,3>::EquationIdVector(EquationIdVectorType& rRes
     }
 }
 
+
+
+template<unsigned int TDim, unsigned int TNumNodes>
+void NavierStokesWallCondition<TDim,TNumNodes>::CalculateLocalSystem(MatrixType& rLeftHandSideMatrix,
+                                      VectorType& rRightHandSideVector,
+                                      ProcessInfo& rCurrentProcessInfo)
+{
+    KRATOS_TRY
+
+    constexpr unsigned int MatrixSize = TNumNodes*(TDim+1);
+    if (rLeftHandSideMatrix.size1() != MatrixSize)
+        rLeftHandSideMatrix.resize(MatrixSize, MatrixSize, false); //false says not to preserve existing storage!!
+    if (rRightHandSideVector.size() != MatrixSize)
+        rRightHandSideVector.resize(MatrixSize, false); //false says not to preserve existing storage!!
+
+    // Struct to pass around the data
+    ConditionDataStruct data;
+    // Allocate memory needed
+    array_1d<double,MatrixSize> rhs_gauss;
+    BoundedMatrix<double,MatrixSize, MatrixSize> lhs_gauss;
+
+    // LHS and RHS contributions initialization
+    noalias(rLeftHandSideMatrix) = ZeroMatrix(MatrixSize,MatrixSize);
+    noalias(rRightHandSideVector) = ZeroVector(MatrixSize);
+    
+    // Compute condition unit normal vector
+    this->CalculateNormal(data.Normal); //this already contains the area
+    const double A = norm_2(data.Normal);
+    data.Normal /= A;
+    
+    // Store the outlet inflow prevention constants in the data structure
+    data.delta = 1e-2; // TODO: Decide if this constant should be fixed or not
+    const ProcessInfo& rProcessInfo = rCurrentProcessInfo; // const to avoid race conditions on data_value_container access/initialization
+    data.charVel = rProcessInfo[CHARACTERISTIC_VELOCITY];
+    
+    // Gauss point information
+    GeometryType& rGeom = this->GetGeometry();
+    const GeometryType::IntegrationPointsArrayType& IntegrationPoints = rGeom.IntegrationPoints(GeometryData::GI_GAUSS_2);
+    const unsigned int NumGauss = IntegrationPoints.size();
+    Vector GaussPtsJDet = ZeroVector(NumGauss);
+    rGeom.DeterminantOfJacobian(GaussPtsJDet, GeometryData::GI_GAUSS_2);
+    const MatrixType Ncontainer = rGeom.ShapeFunctionsValues(GeometryData::GI_GAUSS_2);
+    
+    // Loop on gauss points
+    for(unsigned int igauss = 0; igauss<NumGauss; igauss++)
+    {
+        data.N = row(Ncontainer, igauss);
+        const double J = GaussPtsJDet[igauss];
+        data.wGauss = J * IntegrationPoints[igauss].Weight();
+        ComputeGaussPointRHSContribution(rhs_gauss, data);
+        ComputeGaussPointLHSContribution(lhs_gauss, data);
+        noalias(rLeftHandSideMatrix) += lhs_gauss;
+        noalias(rRightHandSideVector) += rhs_gauss;
+    }
+    
+    // Adding the BEHR2004 contribution if a slip BC is detected
+    // Reference BEHR2004: https://onlinelibrary.wiley.com/doi/abs/10.1002/fld.663
+    if (this->Is(SLIP)){
+
+        Element parent;
```

Comment:
Assuming that a condition will always have the same parent element, we can save a pointer to it and avoid to do the search each time we compute the LHS and RHS contributions. 

@jcotela what you implemented in #3419 stores such parent element in the variable `NEIGHBOUR_ELEMENTS` of each condition. Isn't it? So we can even remove this by calling such functionality once it is merged to the master. Am I right?

Reply:
@rubenzorrilla yes, the idea is that that utility should be able to calculate and store the `NEIGHBOUR_ELEMENTS` condition variable for you (it already calculates it internally, it is just a matter of telling it to save the result), so this calculation does not need to be repeated at the condition level. Also, this avoids the potential for problems with reordering nodes that @philbucher mentioned.